lib/logstash/inputs/exec.rb

```diff
@@ -105,20 +105,24 @@ def execute(queue)
   private
 
   def run_command
-    @io = IO.popen(@command)
-    output = @io.read
-    @io.close # required in order to read $?
-    exit_status = $?.exitstatus
+    @p_in, @p_out, waiter = Open3.popen2(@command)
+    output = @p_out.read
+    exit_status = waiter.value.exitstatus
     [output, exit_status]
   ensure
-    close_io()
+    close_out_and_in
+  end
+
+  def close_out_and_in
+    do_close(@p_out)
+    @p_out = nil
+    do_close(@p_in)
+    @p_in = nil
   end
 
-  # Close @io
-  def close_io
-    return if @io.nil? || @io.closed?
-    @io.close
-    @io = nil
+  def do_close(io)
+    return if io.nil? || io.closed?
+    io.close
   end
```

Comment:
Maybe we can just guard this step against some last minute broken pipes or any other errors from closing an IO object? Something like

```suggestion
  def do_close(io)
    return if io.nil? || io.closed?
    io.close
  rescue => e
    @logger.debug("exception raised while closing io. ignoring..", :io => io, :exception => e.class, :message => e.message)
  end
```